libcst/matchers/_matcher_base.py

```diff
@@ -0,0 +1,638 @@
+        self.options: Sequence[_MatcherT] = actual_options
+
+    def __or__(self, other: _OtherNodeT) -> NoReturn:
+        raise Exception("Cannot use AllOf and OneOf in combination!")
+
+    def __and__(self, other: _OtherNodeT) -> "AllOf[Union[_MatcherT, _OtherNodeT]]":
+        # Without a cast, pyre thinks that the below AllOf is type AllOf[object]
+        # even though it has the types passed into it.
+        return cast(AllOf[Union[_MatcherT, _OtherNodeT]], AllOf(self, other))
+
+    def __invert__(self) -> "OneOf[_MatcherT]":
+        # Invert using De Morgan's Law so we don't have to complicate types.
+        return cast(OneOf[_MatcherT], OneOf(*[DoesNotMatch(m) for m in self.options]))
+
+    def __repr__(self) -> str:
+        return f"AllOf({', '.join([repr(o) for o in self.options])})"
+
+
+class InverseOf(Generic[_MatcherT]):
+    """
+    Matcher node that inverts the match of its child. You can also construct a
+    :class:`InverseOf` matcher by using Python's bitwise invert operator
+    (``~``) with generated matcher classes.
+    """
+
+    def __init__(self, matcher: _MatcherT) -> None:
+        self.matcher = matcher
+
+    def __or__(self, other: _OtherNodeT) -> "OneOf[Union[_MatcherT, _OtherNodeT]]":
+        # Without a cast, pyre thinks that the below OneOf is type OneOf[object]
+        # even though it has the types passed into it.
+        return cast(OneOf[Union[_MatcherT, _OtherNodeT]], OneOf(self, other))
+
+    def __and__(self, other: _OtherNodeT) -> "AllOf[Union[_MatcherT, _OtherNodeT]]":
+        # Without a cast, pyre thinks that the below AllOf is type AllOf[object]
+        # even though it has the types passed into it.
+        return cast(AllOf[Union[_MatcherT, _OtherNodeT]], AllOf(self, other))
+
+    def __getattr__(self, key: str) -> object:
+        # We lie about types to make InverseOf appear transparent. So, its conceivable
+        # that somebody might try to dereference an attribute on the _MatcherT wrapped
+        # node and become surprised that it doesn't work.
+        return getattr(self.matcher, key)
+
+    def __invert__(self) -> _MatcherT:
+        return self.matcher
+
+    def __repr__(self) -> str:
+        return f"DoesNotMatch({repr(self.matcher)})"
+
+
+class MatchIfTrue(Generic[_CallableT]):
+    """
+    Matcher node that matches if its child callable returns ``True``.
+    """
+
+    def __init__(self, func: _CallableT) -> None:
+        # Without a cast, pyre thinks that self.func is not a function, even though
+        # it recognizes that it is a _CallableT bound to Callable.
+        self.func: Callable[..., bool] = cast(Callable[..., bool], func)
+
+    def __or__(
+        self, other: _OtherNodeT
+    ) -> "OneOf[Union[MatchIfTrue[_CallableT], _OtherNodeT]]":
+        # Without a cast, pyre thinks that the below OneOf is type OneOf[object]
+        # even though it has the types passed into it.
+        return cast(
```

Comment:
This is not tested. Can we add a test case for it?
https://codecov.io/gh/Instagram/LibCST/pull/75/diff?src=pr&el=tree#D6-210